hw/dv/verilator/cpp/verilator_memutil.cc

```diff
@@ -187,296 +189,576 @@ bool VerilatorMemUtil::ParseMemArg(std::string mem_argument, std::string &name,
+                           "Run with --meminit=list to get a list.");
+    throw std::runtime_error(oss.str());
+  }
+
+  if (verbose) {
+    std::cout << "Loading data from file `" << filepath << "' into memory `"
+              << name << "'." << std::endl;
+  }
+
+  const MemArea &m = it->second;
 
   switch (type) {
     case kMemImageElf:
-      if (!WriteElfToMem(scope, filepath, size_byte)) {
-        std::cerr << "ERROR: Writing ELF file to memory \"" << m.name << "\" ("
-                  << m.location << ") failed." << std::endl;
-        return false;
-      }
+      WriteElfToMem(m, filepath);
       break;
     case kMemImageVmem:
-      if (!WriteVmemToMem(scope, filepath)) {
-        std::cerr << "ERROR: Writing VMEM file to memory \"" << m.name << "\" ("
-                  << m.location << ") failed." << std::endl;
-        return false;
-      }
+      WriteVmemToMem(m, filepath);
       break;
-    case kMemImageUnknown:
     default:
-      std::cerr << "ERROR: Unknown file type for " << m.name << std::endl;
-      return false;
+      assert(0);
   }
-  return true;
 }
 
-bool VerilatorMemUtil::WriteElfToMem(const svScope &scope,
-                                     const std::string &filepath,
-                                     size_t size_byte) {
-  bool retcode;
-  svScope prev_scope = svSetScope(scope);
+void VerilatorMemUtil::LoadElfToMemories(bool verbose,
+                                         const std::string &filepath) {
+  (void)elf_errno();
+  if (elf_version(EV_CURRENT) == EV_NONE) {
+    throw std::runtime_error(elf_errmsg(-1));
+  }
 
-  uint8_t *buf = nullptr;
-  size_t len_bytes;
+  ElfFile elf(filepath);
 
-  // This "mini buffer" is used to transfer each write to Verilator. It's not
-  // massively efficient, but doing so ensures that we pass 256 bits (32 bytes)
-  // of initialised data each time. This is for simutil_verilator_set_mem
-  // (defined in prim_util_memload.svh), whose "val" argument has SystemVerilog
-  // type bit [255:0].
-  uint8_t minibuf[32];
-  memset(minibuf, 0, sizeof minibuf);
-  assert(size_byte <= sizeof minibuf);
+  size_t file_size;
+  const char *file_data = elf_rawfile(elf.ptr_, &file_size);
+  assert(file_data);
 
-  if (!ElfFileToBinary(filepath, &buf, len_bytes)) {
-    std::cerr << "ERROR: Could not load: " << filepath << std::endl;
-    retcode = false;
-    goto ret;
-  }
-  for (int i = 0; i < (len_bytes + size_byte - 1) / size_byte; ++i) {
-    memcpy(minibuf, &buf[size_byte * i], size_byte);
-    if (!simutil_verilator_set_mem(i, (svBitVecVal *)minibuf)) {
-      std::cerr << "ERROR: Could not set memory byte: " << i * size_byte << "/"
-                << len_bytes << "" << std::endl;
+  size_t phnum = elf.GetPhdrNum();
+  const Elf32_Phdr *phdrs = elf.GetPhdrs();
+
+  for (size_t i = 0; i < phnum; ++i) {
+    const Elf32_Phdr &phdr = phdrs[i];
+    if (phdr.p_type != PT_LOAD)
+      continue;
```

Comment:
For consistency add the message from https://github.com/lowRISC/opentitan/pull/3544/files#diff-bf3d8ae05c262c011ba08283c82ea591R239 ?

Reply:
I decided not to on purpose, because "load an ELF" has pretty well-defined semantics and you wouldn't ever expect to load a debug section or similar. Do you think there's a situation where this might be confusing?

Reply:
True. Did not consider this. It was just confusing to me with switching from meminit to load-elf because the message wasn't there. So I checked what was happening. But going forward this should not be a problem and you make a good point about it being not something anyone would ever expect.

Reply:
Maybe we should update the documentation to explain how the memory can be initialized and how this is happening.

Reply:
Is this documented somewhere? Sorry, I didn't realise, otherwise I would have updated already.

Reply:
We have something in [docs.opentitan](https://docs.opentitan.org/doc/ug/getting_started_verilator/) but nothing specific.
Might be good to have some more details later, explaining what is happening for the different loading options.

Reply:
Ah, right. I've just had a bit of a grep and it looks like this is really just in `getting_started_verilator.md`. We can't change that until we add an address map to the top-level (so there are LMAs for the memory sections). Let's leave it for now, I guess.